programs/gateway/src/lib.rs

```diff
@@ -80,6 +131,106 @@ pub mod gateway {
         Ok(())
     }
 
+    /// Withdraws amount to destination program pda, and calls on_call on destination program
+    ///
+    /// # Arguments
+    /// * `ctx` - The instruction context.
+    /// * `amount` - The amount of SOL to withdraw.
+    /// * `sender` - Sender from ZEVM.
+    /// * `data` - Data to pass to destination program.
+    /// * `signature` - The TSS signature.
+    /// * `recovery_id` - The recovery ID for signature verification.
+    /// * `message_hash` - Message hash for signature verification.
+    /// * `nonce` - The current nonce value.
+    pub fn execute(
+        ctx: Context<Execute>,
+        amount: u64,
+        sender: [u8; 20],
+        data: Vec<u8>,
+        signature: [u8; 64],
+        recovery_id: u8,
+        message_hash: [u8; 32],
+        nonce: u64,
+    ) -> Result<()> {
+        let pda = &mut ctx.accounts.pda;
+
+        if nonce != pda.nonce {
+            msg!(
+                "Mismatch nonce: provided nonce = {}, expected nonce = {}",
+                nonce,
+                pda.nonce,
+            );
+            return err!(Errors::NonceMismatch);
+        }
+
+        let mut concatenated_buffer = Vec::new();
+        concatenated_buffer.extend_from_slice(b"ZETACHAIN");
+        concatenated_buffer.push(InstructionId::Execute as u8);
+        concatenated_buffer.extend_from_slice(&pda.chain_id.to_be_bytes());
+        concatenated_buffer.extend_from_slice(&nonce.to_be_bytes());
+        concatenated_buffer.extend_from_slice(&amount.to_be_bytes());
+        concatenated_buffer.extend_from_slice(&ctx.accounts.destination_program.key().to_bytes());
+        require!(
+            message_hash == hash(&concatenated_buffer[..]).to_bytes(),
+            Errors::MessageHashMismatch
+        );
+
+        msg!("Computed message hash: {:?}", message_hash);
+
+        let address = recover_eth_address(&message_hash, recovery_id, &signature)?;
+        if address != pda.tss_address {
+            msg!("ECDSA signature error");
+            return err!(Errors::TSSAuthenticationFailed);
+        }
```

Comment:
I see we have always this code for withdraw, withdrawSPL, now withdrawAndCall, isn't something that could be facotrized?